io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,394 @@
+
+    thread_.join ();
+
+    pipe_.stop ();
+  }
+
+  void 
+  RealSense2Grabber::start ()
+  {
+    if (isRunning ())
+      return;
+
+    running_ = true;
+
+    rs2::config cfg;
+
+    // capture from file
+    if (file_name_or_serial_number_.rfind (".bag") == file_name_or_serial_number_.length() - 4)
+    {
+      cfg.enable_device_from_file (file_name_or_serial_number_);
+    }
+    else
+    {
+      if (!file_name_or_serial_number_.empty ())
+        cfg.enable_device (file_name_or_serial_number_);
+
+      cfg.enable_stream (RS2_STREAM_COLOR, device_width_, device_height_, RS2_FORMAT_RGB8, target_fps_);
+      cfg.enable_stream (RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_Z16, target_fps_);
+      cfg.enable_stream (RS2_STREAM_INFRARED, device_width_, device_height_, RS2_FORMAT_Y8, target_fps_);
+    }    
+
+    rs2::pipeline_profile prof = pipe_.start (cfg);
+
+    assert (prof.get_stream (RS2_STREAM_COLOR).format () == RS2_FORMAT_RGB8);
+    assert (prof.get_stream (RS2_STREAM_DEPTH).format () == RS2_FORMAT_Z16);
+
+    ir_format_ = prof.get_stream (RS2_STREAM_INFRARED).format (); 
+
+    thread_ = std::thread (&RealSense2Grabber::threadFunction, this);
+
+  }
+
+  void 
+  RealSense2Grabber::stop ()
+  {
+    std::lock_guard<std::mutex> guard (mutex_);
+
+    quit_ = true;
+    running_ = false;
+  }
+
+  bool 
+  RealSense2Grabber::isRunning () const
+  {
+    std::lock_guard<std::mutex> guard (mutex_);
+
+    return running_;
+  }
+
+  float 
+  RealSense2Grabber::getFramesPerSecond () const
+  {
+    return fps_;
+  }
+
+  pcl::PointCloud<pcl::PointXYZ>::Ptr 
+  RealSense2Grabber::convertDepthToPointXYZ (const rs2::points & points)
+  {
+    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZ> ());
+
+    auto sp = points.get_profile ().as<rs2::video_stream_profile> ();
+    cloud->width = sp.width ();
+    cloud->height = sp.height ();
+    cloud->is_dense = false;
+    cloud->points.resize (points.size ());
+
+    auto vertices_ptr = points.get_vertices ();
+
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (int index = 0; index < cloud->points.size (); ++index)
+    {
+      auto ptr = vertices_ptr + index;
+      auto& p = cloud->points[index];
+
+      p.x = ptr->x;
+      p.y = ptr->y;
+      p.z = ptr->z;
+    }
+
+    return cloud;
+  }
+
+  pcl::PointCloud<pcl::PointXYZI>::Ptr 
+  RealSense2Grabber::convertInfraredDepthToPointXYZI (const rs2::points & points, rs2::video_frame & ir)
+  {
+    pcl::PointCloud<pcl::PointXYZI>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZI> ());
+
+    auto sp = points.get_profile ().as<rs2::video_stream_profile> ();
+    cloud->width = sp.width ();
+    cloud->height = sp.height ();
+    cloud->is_dense = false;
+    cloud->points.resize (points.size ());
+
+    auto vertices_ptr = points.get_vertices ();
+    auto texture_ptr = points.get_texture_coordinates ();
+
+    uint8_t r, g, b;
+
+    if (ir_format_ == RS2_FORMAT_UYVY)
```

Comment:
Why repeat the loop two times instead of just applying the conditional branch on the way you fill 'p.intensity`?

Reply:
I was trying to minimize the if call within the large loop.  I've updated it, but we might want to turn it back if we want to increase performance

Reply:
I see... let me think about it.

Reply:
I think we can probably leverage the power lambdas here. Something along the lines of 
```cpp
auto p = (condition)? []() { std::cout << "routine A" << std::endl; } :
                      []() { std::cout << "routine B" << std::endl; } ;

[inside the loop]

p();
```

You'll need to modify of course for the right input argument and the actual functions to be executed.
Don't use captures. Define whatever you need as arguments.

Reply:
This way there is no `if` statement inside the for loop, sure. But compiler will still need to use a jump. You may argue that the condition at least is precomputed, but I'd expect the compiler to be smart enough to figure out `ir_format_ == RS2_FORMAT_UYVY` does not change inside the loop either.